### PR TITLE
[ONNX] Apply Common Subexpression Elimination pass to ONNX export

### LINKED ISSUE #84092: torch.jit.trace doesn't work with autocast on Conv node.
### 🐛 Describe the bug

I was trying a mix precision model, and found that the Conv weight/bias become constant with autocast. So I made a liitle repro, finding that when `torch.jit.trace` is applied with autocast. Trace sanity check would fail.

```python
import torch

class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.cv1 = torch.nn.Conv2d(3, 3, 5, 2, 1)

    def forward(self, x):
        x = self.cv1(x)
        return x

x = torch.randn(10, 3, 20, 20) * 2
m = MyModule().eval()
x = x.cuda()
m = m.cuda()

with torch.no_grad():
    print("outside result: ", torch.jit.trace(m, x))
    with torch.cuda.amp.autocast(enabled = True, dtype=torch.float16):
        print("inside result: ", torch.jit.trace(m, x))
```
Error message
```
ERROR: Graphs differed across invocations!
        Graph diff:
                  graph(%self.1 : __torch__.MyModule,
                        %x : Tensor):
                    %cv1 : __torch__.torch.nn.modules.conv.Conv2d = prim::GetAttr[name="cv1"](%self.1)
                    %4 : int = prim::Constant[value=15]()
                    %5 : bool = prim::Constant[value=1](), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                    %6 : int = prim::Constant[value=0](), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                    %7 : int = prim::Constant[value=1](), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                    %8 : int = prim::Constant[value=2](), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                +   %9 : Tensor = prim::Constant[value=0.01 *  6.7810  6.4636  5.3894 [ CUDAHalfType{3} ]](), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                +   %10 : Tensor = prim::Constant[value=<Tensor>](), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                -   %9 : NoneType = prim::Constant(), scope: __module.cv1
                ?    ^
                +   %11 : NoneType = prim::Constant(), scope: __module.cv1
                ?    ^^
                -   %10 : bool = prim::Constant[value=0](), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                ?     ^
                +   %12 : bool = prim::Constant[value=0](), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                ?     ^
                -   %11 : int = prim::Constant[value=5](), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                ?     ^
                +   %13 : int = prim::Constant[value=5](), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                ?     ^
                -   %weight : Tensor = prim::GetAttr[name="weight"](%cv1)
                -   %bias : Tensor = prim::GetAttr[name="bias"](%cv1)
                -   %14 : Tensor = aten::to(%bias, %11, %10, %10, %9), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                -   %15 : Tensor = aten::to(%weight, %11, %10, %10, %9), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                -   %16 : Tensor = aten::to(%x, %11, %10, %10, %9), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                ?     ^                            --------------
                +   %14 : Tensor = aten::to(%x, %13, %12, %12, %11), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                ?     ^                         +++++++++++++++
                -   %17 : int[] = prim::ListConstruct(%8, %8), scope: __module.cv1
                ?     ^
                +   %15 : int[] = prim::ListConstruct(%8, %8), scope: __module.cv1
                ?     ^
                -   %18 : int[] = prim::ListConstruct(%7, %7), scope: __module.cv1
                ?     ^
                +   %16 : int[] = prim::ListConstruct(%7, %7), scope: __module.cv1
                ?     ^
                -   %19 : int[] = prim::ListConstruct(%7, %7), scope: __module.cv1
                ?     ^
                +   %17 : int[] = prim::ListConstruct(%7, %7), scope: __module.cv1
                ?     ^
                -   %20 : int[] = prim::ListConstruct(%6, %6), scope: __module.cv1
                ?    ^^
                +   %18 : int[] = prim::ListConstruct(%6, %6), scope: __module.cv1
                ?    ^^
                +   %19 : Tensor = aten::_autocast_to_reduced_precision(%14, %5, %12, %13, %4)
                +   %20 : Tensor = aten::_autocast_to_reduced_precision(%10, %5, %12, %13, %4)
                -   %21 : Tensor = aten::_autocast_to_reduced_precision(%16, %5, %10, %11, %4)
                ?                                                        ^^        ^    ^
                +   %21 : Tensor = aten::_autocast_to_reduced_precision(%9, %5, %12, %13, %4)
                ?                                                        ^        ^    ^
                -   %22 : Tensor = aten::_autocast_to_reduced_precision(%15, %5, %10, %11, %4)
                -   %23 : Tensor = aten::_autocast_to_reduced_precision(%14, %5, %10, %11, %4)
                -   %24 : Tensor = aten::_convolution(%21, %22, %23, %17, %18, %19, %10, %20, %7, %10, %10, %5, %5), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                ?     ^                                     ^^   ^^            ---------------      ^    ^
                +   %22 : Tensor = aten::_convolution(%19, %20, %21, %15, %16, %17, %12, %18, %7, %12, %12, %5, %5), scope: __module.cv1 # /home/titaiwang/pytorch/torch/nn/modules/conv.py:459:0
                ?     ^                                ++++++++++     ^^   ^^       +++++           ^    ^
                -   return (%24)
                ?             ^
                +   return (%22)
                ?             ^
        First diverging operator:
        Node diff:
                - %cv1 : __torch__.torch.nn.modules.conv.___torch_mangle_2.Conv2d = prim::GetAttr[name="cv1"](%self.1)
                ?                                                        ^
                + %cv1 : __torch__.torch.nn.modules.conv.___torch_mangle_4.Conv2d = prim::GetAttr[name="cv1"](%self.1)
                ?                                                        ^
```

### Versions

Pytorch nightly

cc @mcarilli @ptrblck

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
## Summary
Exporting graphs with Autocast may fail due to a limitation on JIT tracer. By disabling Autocast cache, tracer works, but there can be performance hit when there is reuse of weights in convolution, for example

By applying CSE, such performance loss can be reverted.

ps: As a comment at #84092 mentioned, disabling Autocast cache is an acceptable workaround and used throughout PyTorch code.

Fixes #84092

### Example: eliminating `%17` and reusing `%16` instead

```python
# BEFORE 
graph(%0 : Float(requires_grad=0, device=cpu)):
  %3 : Scalar = aten::ScalarImplicit(%0), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %13 : int = prim::Constant[value=3](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %14 : int = prim::Constant[value=4](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %15 : int[] = prim::ListConstruct(%13, %14), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %16 : NoneType = prim::Constant(), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %17 : NoneType = prim::Constant(), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %18 : Device = prim::Constant[value="cpu"](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %19 : bool = prim::Constant[value=0](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %20 : Float(3, 4, strides=[4, 1], requires_grad=0, device=cpu) = aten::full(%15, %3, %16, %17, %18, %19), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  return (%20)

AFTER
graph(%0 : Float(requires_grad=0, device=cpu)):
  %3 : Scalar = aten::ScalarImplicit(%0), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %13 : int = prim::Constant[value=3](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %14 : int = prim::Constant[value=4](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %15 : int[] = prim::ListConstruct(%13, %14), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %16 : NoneType = prim::Constant(), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule::
  %18 : Device = prim::Constant[value="cpu"](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %19 : bool = prim::Constant[value=0](), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  %20 : Float(3, 4, strides=[4, 1], requires_grad=0, device=cpu) = aten::full(%15, %3, %16, %16, %18, %19), scope: test_onnx_opset.TestONNXOpset.test_full.<locals>.MyModule:: # /home/thiagofc/dev/github/pytorch/test/onnx/test_onnx_opset.py:347:0
  return (%20)
```

### CODE REVIEW

**Comment by ngimel** [comment]
Lg, lint issue is real?

**Review by ngimel — APPROVED** [review]

**Comment by thiagocrepaldi** [comment]
> Lg, lint issue is real?

yup, missing `def _jit_pass_cse(Graph) -> bool: ...` from `torch/_C/__init__.pyi.in`

**Review by titaiwangms — APPROVED** [review]
Nice finding!

**Review by BowenBao — APPROVED** [review]
LGTM, a few questions and comments.
* To clarify, do we need both disabling caching in autocast and this PR to fully resolve #84092?
* Could you provide a test case demonstrating the effect? I think for duplicated parameters it might have been handled by https://github.com/pytorch/pytorch/blob/4c4e5f6106b69960833d7766799fd4f246aa7cd7/torch/onnx/utils.py#L1498, however this one should cover duplicated constants as well.
* CI failure is just test_operators.py expect files that needs updating due to this change.

**Comment by thiagocrepaldi** [comment]
> LGTM, a few questions and comments.
> 
> * To clarify, do we need both disabling caching in autocast and this PR to fully resolve #84092?
Disabling autocast fix the bug, but might cause a performance decrease due to reused weights, especially in conv nodes.
This PR removes such perf hit by using CSE, and therefore, reusing weights again, instead of creating duplicates

> * Could you provide a test case demonstrating the effect? I think for duplicated parameters it might have been handled by https://github.com/pytorch/pytorch/blob/4c4e5f6106b69960833d7766799fd4f246aa7cd7/torch/onnx/utils.py#L1498
CSE is a little more advanced than just removing existing duplicates; It first generates duplicates through decomposition of complex expressions and then eliminate them

It can decompose expressions to extract a common factor, which will be reused in other places:
```python
# b * c is computed twice here
a = b * c + g;
d = b * c * e;

# After CSE, the same code is rewritten as
tmp = b * c;
a = tmp + g;
d = tmp * e;

# b * c now can be reused in other places, such as `a` and `d`
```

To test this functionality, ideally, we would have to have models with and without CSE and show one has duplicate while the other doesn't. The PR description show such difference because I printed the graph before and after this pass, show the difference is visible. But for a CSE only code, I would have to do a double loop and compare all node against all nodes asserting inequality. Do you think that would be useful?

>   , however this one should cover duplicated constants as well.
> * CI failure is just test_operators.py expect files that needs updating due to this change.
 Yup, just pushed a new update for them
